### PR TITLE
refactor: External ID encoding with lexicographical order

### PR DESCRIPTION
**Pull request by jannden**
Closes: #86 & #87

### CODE REVIEW

File: src/sources/ldap.rs [hunk 1]
```diff
@@ -124,13 +124,11 @@ impl LdapSource {
...
 	attribute: &AttributeMapping,
-	id: &str,
+	_id: &str,
 ) -> Result<String> {
 	match read_search_entry(entry, attribute)? {
 		StringOrBytes::String(entry) => Ok(entry),
-		StringOrBytes::Bytes(_) => {
-			Err(anyhow!("Unacceptable binary value for {} of user `{}`", attribute, id))
-		}
+		StringOrBytes::Bytes(bytes) => Ok(BASE64_STANDARD.encode(&bytes)),
```
Comment (jannden) [inline src/sources/ldap.rs:131]: @tlater-famedly
I couldn't immediately understand why wasn't Bytes accepted?
Also, shall Base64 be sufficient here?

File: src/sources/ldap.rs [hunk 2]
```diff
@@ -90,9 +90,9 @@ impl LdapSource {
 			.any(|flag| status_as_int & flag != 0);
 
 		let ldap_user_id = match read_search_entry(&entry, &self.ldap_config.attributes.user_id)? {
-			// TODO(tlater): Use an encoding that preserves alphabetic order
-			StringOrBytes::Bytes(byte_id) => BASE64_STANDARD.encode(byte_id),
-			StringOrBytes::String(string_id) => BASE64_STANDARD.encode(string_id),
+			// Use hex encoding instead of base64 for consistent alphabetical order
+			StringOrBytes::Bytes(byte_id) => hex::encode(byte_id),
+			StringOrBytes::String(string_id) => hex::encode(string_id.to_lowercase().as_bytes()),
```
Comment (jannden) [inline src/sources/ldap.rs:95]: @emgrav You suggested `base32hex`, was it for a particular reason? I think just a plain `hex` encoding should be sufficient without having to worry about padding.

Comment (emgrav) [reply src/sources/ldap.rs:95]: Yes, regular hex also works. base32hex is slightly more space efficient but that's of no concern here.

File: src/sources/ldap.rs (hunk 1 above)
Comment (tlater-famedly) [reply src/sources/ldap.rs:131]: This function's entire point was throwing an error for non-string values, it's only here to not write the same error handling 6 times x) Hence the doc comment:

```rust
/// Read an an attribute, but assert that it is a string
```

I don't think base64 encoding these values is a good idea, it will be stuff like user names, which should always be strings. We probably should skip syncing users which have non-string-encoded values that we expect to be strings.

This is different for the values that we expect to be raw binary, for those we use the `read_search_entry` function, and handle the non-string case explicitly.

This *is* indeed a change, but I think this just makes more sense. Happy to hear counter arguments.

File: src/sources/ldap.rs (hunk 2 above)
Comment (tlater-famedly) [inline src/sources/ldap.rs:95]: Are (open)ldap IDs not case sensitive?

File: tests/e2e.rs [hunk 3]
```diff
@@ -485,131 +682,135 @@ async fn test_e2e_sync_invalid_phone() {
...
-// 			.expect("could not get user metadata");
-
-// 		assert_eq!(
-// 			preferred_username
-// 				.map(|u| BASE64_STANDARD.decode(u).expect("failed to decode binary attr")),
-// 			Some("validutf8".as_bytes().to_vec())
-// 		);
-// 	}
-// }
+#[test(tokio::test)]
```
Comment (tlater-famedly) [inline tests/e2e.rs:685]: Yep, hence these tests were commented out; I think we should not support this use case anymore.

We *should* test that binary user IDs in fact work. Maybe we can change the attribute we use for the ID to `userSMIMECertificate` with the config?

Comment (jannden) [reply tests/e2e.rs:684]: I have created a new test and updated the two old.

File: src/sources/ldap.rs (hunk 1 above)
Comment (jannden) [reply src/sources/ldap.rs]: Okay, reverted.

File: src/sources/ldap.rs (hunk 2 above)
Comment (jannden) [reply src/sources/ldap.rs:94]: Not sure about this. Let's discuss.

Comment (tlater-famedly) [reply src/sources/ldap.rs:94]: Well, if they are case sensitive this will cause collisions, while if it they are case sensitive there's a chance that they might change the casing randomly.

I assume the latter is probably not the case, and we should just remove the `to_lowercase()`. Is there a particular reason we do that?

Comment (jannden) [reply src/sources/ldap.rs:94]: Since there is not a particular reason, I agree with you - it's logical to remove the lowercase conversion.

File: src/bin/migrate.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
+//! This binary is used to migrate user IDs from base64 to hex encoding.
```
Comment (tlater-famedly) [inline src/bin/migrate.rs:1]: Can we do this in a separate PR so we get the base functionality ready for a test run quicker?

**Review by tlater-famedly — APPROVED** [review]
